node/core/candidate-validation/src/lib.rs

```diff
@@ -296,30 +302,53 @@ where
 	Context: SubsystemContext<Message = CandidateValidationMessage>,
 	Context: overseer::SubsystemContext<Message = CandidateValidationMessage>,
 {
-	let (validation_data, validation_code) =
-		match find_assumed_validation_data(ctx, &descriptor).await? {
-			AssumptionCheckOutcome::Matches(validation_data, validation_code) =>
-				(validation_data, validation_code),
-			AssumptionCheckOutcome::DoesNotMatch => {
-				// If neither the assumption of the occupied core having the para included or the assumption
-				// of the occupied core timing out are valid, then the persisted_validation_data_hash in the descriptor
-				// is not based on the relay parent and is thus invalid.
-				return Ok(Ok(ValidationResult::Invalid(InvalidCandidate::BadParent)))
-			},
-			AssumptionCheckOutcome::BadRequest =>
-				return Ok(Err(ValidationFailed("Assumption Check: Bad request".into()))),
-		};
+	let validation_data = match find_assumed_validation_data(ctx, &descriptor).await? {
+		AssumptionCheckOutcome::Matches(validation_data) => validation_data,
+		AssumptionCheckOutcome::DoesNotMatch => {
+			// If neither the assumption of the occupied core having the para included or the assumption
+			// of the occupied core timing out are valid, then the persisted_validation_data_hash in the descriptor
+			// is not based on the relay parent and is thus invalid.
+			return Ok(Ok(ValidationResult::Invalid(InvalidCandidate::BadParent)))
+		},
+		AssumptionCheckOutcome::BadRequest =>
+			return Ok(Err(ValidationFailed("Assumption Check: Bad request".into()))),
+	};
 
-	let validation_result = validate_candidate_exhaustive(
-		validation_host,
-		validation_data,
-		validation_code,
+	let mut validation_result = validate_candidate_exhaustive(
+		&mut validation_host.clone(),
+		validation_data.clone(),
+		None,
 		descriptor.clone(),
-		pov,
+		pov.clone(),
 		metrics,
 	)
 	.await;
 
+	// In case preimage for the supplied code hash was not found by the
+	// validation host, request the code from Runtime API and try again.
+	if let Ok(Err(ref err)) = validation_result {
+		if err.0.as_str() == "Code not found" {
```

Comment:
No, you cannot do that. I perceive this unreliable. At the very least you can do an enum or otherwise make sure the compiler helps you to avoid a disaster.

I think we can move the code query into the `validate_candidate_exhaustive`. Specifically, if the caller didn't pass preimage, then if `validation_backend.validate_candidate` returned `ArtifactNotFound` it would ask the runtime for the validation code by hash and retry. This can be abstracted as a function.

Reply:
I agree here, it's just that the error holds a string in the first place. That's a good hint to handle an error inside of the exhaustive function

Reply:
> I agree here, it's just that the error holds a string in the first place. 

Yeah but I think it's fine. The error has multiple purposes: reporting errors to humans and signalling a condition to trigger some recovery logic. This type was used only for reporting errors and a string could do, but IMO a string is an inappropriate tool for matching as a condition. It sometimes happens that new requirements can lead to the need on matching and in that case we should do a refactoring. Only in very rare and extreme cases we can reach to string matching counting on eventual refactoring. This particular case is far from that though.